File: 0206-reverse-linked-list/0206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Initialize previous pointer to None
        prev = None
        current = head
        
        while current:
            # Save the next node
            next_node = current.next
            # Reverse the current node's pointer
            current.next = prev
            # Move the previous and current pointers one step forward
            prev = current
            current = next_node
        
        # At the end, prev will be the new head of the reversed list
        return prev

File: A. Easy/0206-reverse-linked-list/0206-reverse-linked-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* reverseList(struct ListNode* head) {
    // Initialize previous pointer to NULL since the new head's next will be NULL
    struct ListNode *prev = NULL;
    // Initialize current pointer to the head of the list
    struct ListNode *current = head;
    // Initialize next pointer to NULL
    struct ListNode* nextNode = NULL;

    // Traverse the list
    while (current != NULL) {
        // Store the next node before changing the link
         nextNode = current->next;
        // Reverse the current node's pointer to point to the previous node
        current->next = prev;
        // Move the previous pointer forward to the current node
        prev = current;
        // Move to the next node in the original list
        current = nextNode;
    }

    // After the loop, prev will point to the new head of the reversed list
    return prev;
}